src/common/commonutils/CommandUtils.c

```diff
@@ -29,270 +31,200 @@ static int NormalizeStatus(int status)
-{
-    const int callbackIntervalSeconds = 5; //seconds
-    const int defaultCommandTimeout = 60; //seconds
+#define BUFFER_SIZE 1024
 
-    pid_t intermediateProcess = -1;
-    pid_t workerProcess = -1;
-    pid_t timerProcess = -1;
-    pid_t childProcess = -1;
-    int status = -1;
-    int intermediateStatus = -1;
-    int totalWaitSeconds = 0;
-    int timeout = (timeoutSeconds > 0) ? timeoutSeconds : defaultCommandTimeout;
+int ExecuteCommand(void* context, const char* command, bool replaceEol, bool forJson, unsigned int maxTextResultBytes, unsigned int timeoutSeconds,
+    char** textResult, CommandCallback callback, OsConfigLogHandle log)
+{
+    int workerPid = -1;
+    int pipefd[2] = {0};
 
-    bool mainProcessThread = (bool)(getpid() == gettid());
+    if (NULL == command)
+    {
+        OsConfigLogDebug(log, "Command cannot be NULL");
+        return -1;
+    }
+    // Create a pipe, then fork. Forked process dup2()s the write pipe end to stdout and stderr,
+    // then execl()s the shell with the given command. The main process uses timeoutable select()
+    // to read from the pipe, and keep track of both command  timeout and callbacks. The read loop
+    // ends when the read() returns EOF.
 
-    fflush(NULL);
+    if (0 != pipe(pipefd))
+    {
+        OsConfigLogError(log, "Cannot create pipe for command '%s' - %s", command, strerror(errno));
+        return -1;
+    }
 
-    if ((timeoutSeconds > 0) || (NULL != callback))
+    workerPid = fork();
+    if (workerPid < 0)
     {
-        OsConfigLogDebug(log, "SystemCommand: executing command '%s' with timeout of %d seconds and%scancelation on %s thread",
-            command, timeout, (NULL == callback) ? " no " : " ", mainProcessThread ? "main process" : "worker");
+        OsConfigLogError(log, "Cannot fork for command '%s' - %s", command, strerror(errno));
+        close(pipefd[0]);
+        close(pipefd[1]);
+        return -1;
+    }
 
-        // Fork an intermediate process to act as the parent for two more forked processes:
-        // one to actually execute the system command and the other to sleep and act as a timer.
-        // Whichever of these two children processes finishes first, that causes the other to be killed.
-        // The intermediate process exists as a parent for these (instead of the current process being
-        // the parent) to avoid collision with any other children processes.
+    if (0 == workerPid)
+    {
+        // Child process
+        close(pipefd[0]);
+        if (STDOUT_FILENO != dup2(pipefd[1], STDOUT_FILENO))
+        {
+            exit(-1);
+        }
+        if (STDERR_FILENO != dup2(pipefd[1], STDERR_FILENO))
+        {
+            exit(-1);
+        }
+        execl("/bin/sh", "sh", "-c", command, (char*)NULL);
+        // If execl() fails, exit with the error code
+        exit(errno);
+    }
+    else
+    {
+        // Main process
```

Comment:
>// Main process [](http://example.com/codeflow?start=8&length=15)

This comment could be misleading. The main process is the one that calls this function. The forked process is the worker process that executes the command. Which one is here? May want to clarify in comment.

Reply:
As you stated - this is the main process, the one that calls this function.